skorch/callbacks/training.py

```diff
@@ -273,3 +278,173 @@ def _sink(self, text, verbose):
+    >>> cb = ParamMapper('linear0.weight', at=at, fn=init)
+    >>> net = Net(myModule, callbacks=[cb])
+
+    Periodically freeze and unfreeze all embedding layers:
+
+    >>> def my_sched(net, _at, _fn):
+    ...    if len(net.history) % 2 == 0:
+    ...        return skorch.utils.freeze_parameter
+    ...    else:
+    ...        return skorch.utils.unfreeze_parameter
+    >>> cb = ParamMapper('embedding*.weight', schedule=my_sched)
+    >>> net = Net(myModule, callbacks=[cb])
+
+    Parameters
+    ----------
+    patterns : str or callable or list
+      The pattern(s) to match parameter names against. Patterns are
+      UNIX globbing patterns as understood by :func:`~fnmatch.fnmatch`.
+      Patterns can also be callables which will get called with the
+      parameter name and are regarded as a match when the callable
+      returns a truthy value.
+
+      This parameter also supports lists of str or callables so that
+      one ``ParamMapper`` can match a group of parameters.
+
+      Example: ``'linear*.weight'`` or ``['linear0.*', 'linear1.bias']``
+      or ``lambda name: name.startswith('linear')``.
+
+    fn : function
+      The function to apply to each parameter separately.
+
+    at : int or callable
+      In case you specify an integer it represents the epoch number the
+      function ``fn`` is applied to the parameters, in case ``at`` is
+      a function it will receive ``net`` as parameter and the function
+      is applied to the parameter once ``at`` returns ``True``.
+
+    schedule : callable or None
+      If specified this callable supersedes the static ``at``/``fn``
+      combination by dynamically returning the function that is applied
+      on the matched parameters. This way you can, for example, create a
+      schedule that periodically freezes and unfreezes layers.
+
+      The callable's signature is ``schedule(net: NeuralNet) -> callable``.
+
+    """
+    def __init__(self, patterns, fn=noop, at=1, schedule=None):
+        self.at = at
+        self.fn = fn
+        self.schedule = schedule
+        self.patterns = patterns
+
+    def initialize(self):
+        if not self.schedule:
+            self.schedule = self._default_schedule
+
+        if not isinstance(self.patterns, (list, tuple)):
+            self.patterns = [self.patterns]
+
+        if isinstance(self.at, int):
+            if self.at <= 0:
+                raise ValueError(
+                    'Invalid value for `at` (at={}). The first possible '
+                    'epoch number is 1.'.format(self.at))
+            self.at = partial(self._epoch_at, epoch=self.at)
+
+        return self
+
+    def named_parameters(self, net):
+        return list(net.module_.named_parameters())
+
+    def filter_parameters(self, patterns, params):
+        for pattern in patterns:
+            pattern_fn = lambda n: fnmatch(n, pattern)
+            pattern_fn = pattern if callable(pattern) else pattern_fn
```

Comment:
Maybe `pattern_fn = partial(fnmatch, pat=pattern) if callable(pattern) else pattern` is more elegant?

Reply:
Also, assigning lambda can be dangerous (though here it should be fine):
```
>>> x = 3
>>> f = lambda y: x + y
>>> g = partial(add, x)
>>> x = 10
>>> f(5)
15
>>> g(5)
8
```